collection_mutation.cc

```diff
@@ -309,15 +308,35 @@ deserialize_collection_mutation(const abstract_type& type, bytes_view in) {
     }
     auto nr = read_simple<uint32_t>(in);
     ret.cells.reserve(nr);
-    for (uint32_t i = 0; i != nr; ++i) {
-        // FIXME: we could probably avoid the need for size
-        auto ksize = read_simple<uint32_t>(in);
-        auto key = read_simple_bytes(in, ksize);
-        auto vsize = read_simple<uint32_t>(in);
-        // value_comparator(), ugh
-        auto value = atomic_cell_view::from_bytes(ctype.value_comparator()->imr_state().type_info(), read_simple_bytes(in, vsize));
-        ret.cells.emplace_back(key, value);
-    }
+
+    visit(type, make_visitor(
+        [&] (const collection_type_impl& ctype) {
+            // value_comparator(), ugh
+            auto& type_info = ctype.value_comparator()->imr_state().type_info();
+            for (uint32_t i = 0; i != nr; ++i) {
```

Comment:
This loop and the one for UDTs look very similar. It could make sense to generalize them and just pass lambda that reads key and value in collection/udt specific way.